graphula-core/src/Graphula.hs

```diff
@@ -167,7 +178,15 @@ instance (Monad m, MonadGraphulaFrontend m) => MonadGraphulaFrontend (GraphulaT
 
 newtype GraphulaIdempotentT m a =
   GraphulaIdempotentT {runGraphulaIdempotentT' :: ReaderT (IORef (m ())) m a}
-  deriving (Functor, Applicative, Monad, MonadThrow, MonadCatch, MonadMask, MonadIO, MonadReader (IORef (m ())))
+  deriving (Functor, Applicative, Monad, MonadIO, MonadReader (IORef (m ())))
+
+instance MonadUnliftIO m => MonadUnliftIO (GraphulaIdempotentT m) where
+  {-# INLINE askUnliftIO #-}
+  askUnliftIO = GraphulaIdempotentT $ withUnliftIO $ \u ->
+    return $ UnliftIO $ unliftIO u . runGraphulaIdempotentT'
+  {-# INLINE withRunInIO #-}
+  withRunInIO inner = GraphulaIdempotentT $ withRunInIO $ \run ->
+    inner $ run . runGraphulaIdempotentT'
```

Comment:
FWIW: I just copied the `ReaderT` implementation and followed the type errors. I totally don't understand how all this works.

Reply:
Yeah, this is essentially a `newtype` deriving of this instance. Could you just `newtype` derive this?

Reply:
No, GHC spat an incomprehensible error when I tried that.